File: apps/api/src/repositories/savedJob.repository.ts
import { SavedJobModel } from "@believe-ai/server";
import type { Job } from "@believe-ai/shared";

function toSnapshot(job: Omit<Job, "id" | "postedBy" | "createdAt" | "updatedAt" | "isSaved">) {
  return {
    source: job.source,
    title: job.title,
    company: job.company,
    location: job.location,
    description: job.description,
    skills: job.skills,
    employmentType: job.employmentType,
    workMode: job.workMode,
    experienceLevel: job.experienceLevel,
    salaryMin: job.salaryMin,
    salaryMax: job.salaryMax,
    recruiterLinkedIn: job.recruiterLinkedIn,
    applyUrl: job.applyUrl,
  };
}

export const savedJobRepository = {
  /** Ids of every job this user has saved — used to annotate search results with isSaved. */
  async savedJobIds(userId: string): Promise<Set<string>> {
    const rows = await SavedJobModel.find({ userId }, { jobId: 1 });
    return new Set(rows.map((r) => r.jobId));
  },

  find(userId: string, jobId: string) {
    return SavedJobModel.findOne({ userId, jobId });
  },

  save(userId: string, jobId: string, job: Omit<Job, "id" | "postedBy" | "createdAt" | "updatedAt" | "isSaved">) {
    return SavedJobModel.create({ userId, jobId, snapshot: toSnapshot(job) });
  },

  unsave(userId: string, jobId: string) {
    return SavedJobModel.deleteOne({ userId, jobId });
  },

  listByUser(userId: string) {
    return SavedJobModel.find({ userId }).sort({ createdAt: -1 });
  },
};

File: apps/backend/repositories/saved_job_repository.py
"""Beanie-backed saved-job access — mirrors apps/api's savedJob.repository.ts."""

from __future__ import annotations

from bson import ObjectId

from models.saved_job import JobSnapshot, SavedJob


async def saved_job_ids(user_id: ObjectId) -> set[str]:
    """Ids of every job this user has saved — used to annotate search results with isSaved."""
    docs = await SavedJob.find(SavedJob.userId == user_id).to_list()
    return {doc.jobId for doc in docs}


async def find(user_id: ObjectId, job_id: str) -> SavedJob | None:
    return await SavedJob.find_one(SavedJob.userId == user_id, SavedJob.jobId == job_id)


async def save(user_id: ObjectId, job_id: str, snapshot: JobSnapshot) -> SavedJob:
    doc = SavedJob(userId=user_id, jobId=job_id, snapshot=snapshot)
    await doc.insert()
    return doc


async def unsave(user_id: ObjectId, job_id: str) -> None:
    await SavedJob.find(SavedJob.userId == user_id, SavedJob.jobId == job_id).delete()


async def list_by_user(user_id: ObjectId) -> list[SavedJob]:
    return await SavedJob.find(SavedJob.userId == user_id).sort("-createdAt").to_list()
